scanner/cmd/scanner/main.go

```diff
@@ -195,18 +197,43 @@ func createBackends(ctx context.Context, cfg *config.Config) (*Backends, error)
 	return &b, nil
 }
 
-// Close closes all the backends used by the scanner.
-func (b *Backends) Close(ctx context.Context) {
+// getServices returns the list of the API services based on the backends provided.
+func getServices(b *Backends) []grpc.APIService {
+	var srvs []grpc.APIService
+	if b.Indexer != nil {
+		srvs = append(srvs, services.NewIndexerService(b.Indexer))
+	}
 	if b.Matcher != nil {
-		err := b.Matcher.Close(ctx)
-		if err != nil {
-			zlog.Error(ctx).Err(err).Msg("closing matcher")
+		// Set the index report getter to the remote indexer if available, otherwise the
+		// local indexer. A nil getter is ok, see implementation.
+		var getter indexer.ReportGetter
+		getter = b.Indexer
+		if b.RemoteIndexer != nil {
+			getter = b.RemoteIndexer
 		}
+		srvs = append(srvs, services.NewMatcherService(b.Matcher, getter))
 	}
-	if b.Indexer != nil {
-		err := b.Indexer.Close(ctx)
-		if err != nil {
-			zlog.Error(ctx).Err(err).Msg("closing indexer")
+	return srvs
+}
+
+// Close closes all the backends used by the scanner.
+func (b *Backends) Close(ctx context.Context) {
+	type Closeable interface {
+		Close(context.Context) error
+	}
+	for _, backend := range []struct {
+		name     string
+		instance Closeable
+	}{
+		{"matcher", b.Matcher},
+		{"indexer", b.Indexer},
+		{"remote indexer", b.RemoteIndexer},
+	} {
+		if backend.instance != nil {
+			err := backend.instance.Close(ctx)
+			if err != nil {
+				zlog.Error(ctx).Err(err).Msgf("closing %s", backend.name)
+			}
```

Comment:
nit: `err` is not needed after this and I don't think the conditional is terribly long

```
if err := backend.instance.Close(ctx); err != nil {
        zlog.Error(ctx).Err(err).Msgf("closing %s", backend.name)
}
```

Reply:
I think it is slightly easier to read in two separate statements.

Reply:
Please close if you don't have any other comments. Thanks.